tensorflow/core/profiler/internal/gpu/rocm_tracer.h

```diff
@@ -0,0 +1,302 @@
+      std::numeric_limits<uint32_t>::max();
+  static constexpr uint32 kInvalidThreadId =
+      std::numeric_limits<uint32_t>::max();
+  static constexpr uint32 kInvalidCorrelationId =
+      std::numeric_limits<uint32_t>::max();
+  static constexpr uint64 kInvalidStreamId =
+      std::numeric_limits<uint64_t>::max();
+  RocmTracerEventType type;
+  RocmTracerEventSource source;
+  RocmTracerEventDomain domain;
+  std::string name;
+  // This points to strings in AnnotationMap, which should outlive the point
+  // where serialization happens.
+  absl::string_view annotation;
+  uint64 start_time_ns;
+  uint64 end_time_ns;
+  uint32 device_id = kInvalidDeviceId;
+  uint32 correlation_id = kInvalidCorrelationId;
+  uint32 thread_id = kInvalidThreadId;
+  int64 stream_id = kInvalidStreamId;
+  union {
+    MemcpyDetails memcpy_info;      // If type == Memcpy*
+    MemsetDetails memset_info;      // If type == Memset*
+    MemAllocDetails memalloc_info;  // If type == MemoryAlloc
+    KernelDetails kernel_info;      // If type == Kernel
+  };
+};
+
+void DumpRocmTracerEvent(const RocmTracerEvent& event, uint64 start_walltime_ns,
+                         uint64 start_gputime_ns);
+
+struct RocmTracerOptions {
+  // map of domain --> ops for which we need to enable the API callbacks
+  // If the ops vector is empty, then enable API callbacks for entire domain
+  std::map<activity_domain_t, std::vector<uint32_t> > api_callbacks;
+
+  // map of domain --> ops for which we need to enable the Activity records
+  // If the ops vector is empty, then enable Activity records for entire domain
+  std::map<activity_domain_t, std::vector<uint32_t> > activity_tracing;
+};
+
+struct RocmTraceCollectorOptions {
+  // Maximum number of events to collect from callback API; if -1, no limit.
+  // if 0, the callback API is enabled to build a correlation map, but no
+  // events are collected.
+  uint64 max_callback_api_events;
+  // Maximum number of events to collect from activity API; if -1, no limit.
+  uint64 max_activity_api_events;
+  // Maximum number of annotation strings that we can accommodate.
+  uint64 max_annotation_strings;
+  // Number of GPUs involved.
+  uint32 num_gpus;
+};
+
+class AnnotationMap {
+ public:
+  explicit AnnotationMap(uint64 max_size) : max_size_(max_size) {}
+  void Add(uint32 correlation_id, const std::string& annotation);
+  absl::string_view LookUp(uint32 correlation_id);
+
+ private:
+  struct AnnotationMapImpl {
+    // The population/consumption of annotations might happen from multiple
+    // callback/activity api related threads.
+    absl::Mutex mutex;
+    // Annotation tends to be repetitive, use a hash_set to store the strings,
+    // an use the reference to the string in the map.
+    absl::node_hash_set<std::string> annotations;
+    absl::flat_hash_map<uint32, absl::string_view> correlation_map;
+  };
+  const uint64 max_size_;
+  AnnotationMapImpl map_;
+
+  TF_DISALLOW_COPY_AND_ASSIGN(AnnotationMap);
```

Comment:
We actually discourage TF_DISALLOW_COPY_AND_ASSIGN for readability reasons. For this and below, could you move the deleted interface within public accessor scope?

i.e.
```suggestion
 public:
  // Not copyable or movable
  AnnotationMap(const AnnotationMap&) = delete;
  AnnotationMap& operator=(const AnnotationMap&) = delete;
```

See this for example: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/profiler/rpc/client/profiler_client.h;l=53

Reply:
this will be fixed in the next commit/push